hera_sim/simulate.py

```diff
@@ -0,0 +1,335 @@
+                method = ""
+                func(obj, **kwargs)
+
+            msg = "\nhera_sim v{version}: Added {component} {method_name}with kwargs: {kwargs}"
+
+            obj.data.history += msg.format(
+                version=version,
+                component="".join(name.split("_")[1:]),
+                method_name=method,
+                kwargs=kwargs,
+            )
+
+        return new_func
+
+
+class Simulator:
+    """
+    Primary interface object for hera_sim.
+
+    Produces visibility simulations with various independent sky- and instrumental-effects, and offers the resulting
+    visibilities in :class:`UVData` format
+    """
+
+    def __init__(
+        self,
+        data_filename=None,
+        n_freq=None,
+        n_times=None,
+        antennas=None,
+        ant_pairs=None,
+        pols=None,
+        time_per_integ=10.7,
+        min_freq=0.1,
+        channel_bw=0.1 / 1024.0,
+        instrument="hera_sim",
+        telescope_location=None,
+        telescope_lat_lon_alt=None,
+        object_name="sim_data",
+        start_jd=2458119.5,
+        vis_units="uncalib",
+        **kwargs
+    ):
+
+        self.data_filename = data_filename
+
+        if self.data_filename is None:
+            # Create an empty UVData object.
+
+            # Ensure required parameters have been set.
+            assert (
+                n_freq is not None
+            ), "if data_filename not given, n_freq must be given"
+            assert (
+                n_times is not None
+            ), "if data_filename not given, n_times must be given"
+            assert (
+                antennas is not None
+            ), "if data_filename not given, antennas must be given"
+            assert (
+                ant_pairs is not None
+            ), "if data_filename not given, ant_pairs must be given"
+
+            # Default values for mutable parameters
+            if pols is None:
+                pols = ["xx"]
+            if telescope_location is None:
+                telescope_location = io.HERA_LOCATION
+            if telescope_lat_lon_alt is None:
+                telescope_lat_lon_alt = io.HERA_LAT_LON_ALT
+
+            # Actually create it
+            self.data = io.empty_uvdata(
+                nfreq=n_freq,
+                ntimes=n_times,
+                ants=antennas,
+                antpairs=ant_pairs,
+                pols=pols,
+                time_per_integ=time_per_integ,
+                min_freq=min_freq,
+                channel_bw=channel_bw,
+                instrument=instrument,
+                telescope_location=telescope_location,
+                telescope_lat_lon_alt=telescope_lat_lon_alt,
+                object_name=object_name,
+                start_jd=start_jd,
+                vis_units=vis_units,
+            )
+
+        else:
+            self.data = self._read_data(self.data_filename, pols, **kwargs)
+
+    def _read_data(self, filename, polarizations, **kwargs):
```

Comment:
why is `polarizations` special? It is a `kwarg` of `uv.read` so you can just lump it in with `**kwargs`